### PR TITLE
[CPP Graph] Fused Attention Doc

### PR DESCRIPTION
**Pull request by DDEle**
## Description
Add doc for jblas-based fused attention.

## Expected Behavior & Potential Risk
N/A

## How has this PR been tested?
Spell check CI

## Dependency Change?
No

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+Fused Attention
+=======
+
+Attention (including MHA, GQA, MHA) is one of the key parts of transformers and also the bottleneck in many scenarios. To implement various optimizations, a fused attention layer and corresponding utilities for the customized KV-cache it uses are introduced.
```
Comment (hshen14) [inline intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:4]: and also the performance critical in many scenarios.

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+Note, this doc assumes you have the basic knowledge of this cpp graph implementation including `ne_tensor::ne`, `ne_tensor::nb` etc.
+
+## KV-cache initialization
+The memory for kv-cache is allocated in `/models/model_utils/model_utils.cpp`. As the fused attention implementation requires certain instruction extensions and potentially some other limitations, fused attention is enabled only if `jblas_reordered_attn_fp32_support()` aggress, denoting with `memory_type = NE_TYPE_JBLAS`. Next, `get_batch_kv_elements_from_gpt_params()` will give the sizes (in terms of bytes if fused attention enabled, or in terms elements if fused attention disabled) of k-cache and v-cache respectively for each batch and each layer. The KV-cache is finally prepared with these 2 sizes by creating `ne_new_tensor` inside `model.kv_self.k/.v` or `model.layer[il].k_cache/.v_cache`.
+
+## KV-cache Append
+KV-cache is appended every time a new pair of K and V are generated by evaluating inner product for QKV (ne_mul_qkv). This operation append an additional K/V-tensor on the dimension of `n_past` (or the dimension of sequence length) (i.e. resulting `n_past = n_past + N`).
```
Comment (hshen14) [inline intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:12]: Where is the definition of N?

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 1 above)
Comment (hshen14) [inline intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:4]: bottleneck -> performance critical

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+> Note: Currently the K and V tensor to append must be contiguous in the dimension of head size (i.e. `head_dim` in some implementation). The strides (`ne_tensor::nb`) of the other 3 dimensions are configurable.
+
+## Fused Attention Computation
+With the KV-cache of customized type and layout, the attention can be computed at its best performance. `ne_flash_attn` accepts a Q-tensor of `batch x head_num x N x head_size` and outputs a result tensor of `batch x N x head_num x head_size` (totally contiguous).
+
+> Node: similar to the append operation, Q-tensor must be contiguous in the dimension of head and its strides (`ne_tensor::nb`) of the other 3 dimensions are configurable.
+
+## More KV-cache Operations is Being Implemented...
+While we are working on extra KV-cache operations, one can simply parallel the whole kv-cache operations and fused attention on commonly-parallelizable dimensions. Just pass each part to every KV-cache operations (and join them together if needed).
```
Comment (hshen14) [inline intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:24]: join -> concatenate?

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:24]: Well, I was trying to refer to the general merging operation (just like the reduce step of the MapReduce programming model).

For example, one may try to dispatch the first half of the seqlen of K/V-cache to one node and the second half to another, both computing a partial result of MHA (but of the full shape), and combining them together as long as some softmax statistics of the two halves are kept (referring the $𝑇𝑐$ loop, $l_i$ and $m_i$ of the $\mbox{Algorithm 1}$ of [the flash attention paper](https://arxiv.org/pdf/2205.14135.pdf)).

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 2 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:12]: `N` is the name of a parameter of the "mode_eval_internal" function of every model defined in `/model/xxx/xxx.cpp`. It should be the number of tokens in the first-token inference and be 1 in the next token inference. I assumed the readers of this doc are familiar with our cpp graph code. I'm not sure how to precisely define it without too much verbosity. Any suggestions?

Comment (hshen14) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:12]: Doc should be self-contained. How about N is the length of input tokens for the first token generation and always 1 for the next token generation?

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
+Fused Attention
+=======
+
```
Comment (kevinintel) [inline intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:3]: please add link of the code

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 2 above)
Comment (kevinintel) [inline intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:12]: you can give the code snippet

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+The tensor of K and V should be both permuted (with `ne_permute`) to `batch x N x head_num x head_size`. The permuted K/V tensors can then be passed to `ne_flash_attn_update_k()/_v()`, together with `n_past` as an offset in the dimension of sequence length, to concatenate to a "view" of the cached K/V of the current layer and current batch.
+
+> Note: Currently the K and V tensor to append must be contiguous in the dimension of head size (i.e. `head_dim` in some implementation). The strides (`ne_tensor::nb`) of the other 3 dimensions are configurable.
+
+## Fused Attention Computation
+With the KV-cache of customized type and layout, the attention can be computed at its best performance. `ne_flash_attn` accepts a Q-tensor of `batch x head_num x N x head_size` and outputs a result tensor of `batch x N x head_num x head_size` (totally contiguous).
+
+> Node: similar to the append operation, Q-tensor must be contiguous in the dimension of head and its strides (`ne_tensor::nb`) of the other 3 dimensions are configurable.
+
+## More KV-cache Operations is Being Implemented...
```
Comment (kevinintel) [inline intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:23]: please add support matrix

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+## KV-cache Append
+KV-cache is appended every time a new pair of K and V are generated by evaluating inner product for QKV (ne_mul_qkv). This operation append an additional K/V-tensor on the dimension of `n_past` (or the dimension of sequence length) (i.e. resulting `n_past = n_past + N`).
+
+The tensor of K and V should be both permuted (with `ne_permute`) to `batch x N x head_num x head_size`. The permuted K/V tensors can then be passed to `ne_flash_attn_update_k()/_v()`, together with `n_past` as an offset in the dimension of sequence length, to concatenate to a "view" of the cached K/V of the current layer and current batch.
```
[the remark this answers is not in the document]
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:14]: It is recommended to implement TP with suggestions in the "Tips for parallelism" section.

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 5 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md]: Added in 4f7f225

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 3 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md]: clarified in 4f7f225

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 1 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md]: Good idea! Updated in 4f7f225

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 2 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md]: I discussed with @a32543254 and we find that using the naming of the [streaming-llm paper](https://arxiv.org/abs/2309.17453) seems to be good idea. Updated in 4f7f225.

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md]: Note sure if putting code for KV-cache updating here would help. For example, `ne_view_4d` and `ne_build_forward_expand` is hard to understand for readers who are not familiar with cpp graph code (or ggml code).

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 4 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:3]: Good idea! Added in 96373ff

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
+Fused Attention
+=======
+
+Attention (including MHA, GQA, MHA) is one of the key parts of transformers and also the performance critical in many scenarios. To implement various optimizations, a fused attention layer and corresponding utilities for the customized KV-cache it uses are introduced. As an example, fused attention can reduce the cost of MHA from 8521.276 ms (17.044 ms) to 248.586 ms (7.944 ms) of a 1975-token llama-7b first-token inference[[1]](#1).
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:4]: `including MHA, GQA, MQA`?

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+model_print_timings:       total time = 11175.74 ms
+========== eval time log of each prediction ==========
+prediction   0, time: 9748.26ms
+prediction   1, time: 50.56ms
+prediction   2, time: 50.00ms
+prediction   3, time: 50.37ms
+```
+
+Where:
+|                  | pred 0 fused attn disabled | pred 0 fused attn enabled | pred 3 fused attn disabled | pred 3 fused attn enabled |
```
Comment (kevinintel) [inline intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:225]: users don't know what's pred0, you can use 1st token

**Comment by airMeng** [comment]
@intellinjun @CeciliaWwq @zhenwei-intel @Zhenzhong1 Since MHA is quite vital for all coming features like TP and streaming, model owners please have a look and raise any questions you have.

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 8 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md]: Updated in c4179aa

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 7 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md]: Corrected in c4179aa

File: intel_extension_for_transformers/llm/runtime/graph/fused_attention.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:14]: It you want the developer of fused-attn to implement TP inside, then it is not necessary to worry about the input & processing of input of fused-attn-layer neither. Again: ne_permute will not do anything to the data of a tensor during graph_compute.

[the remark this answers is not in the document]
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/fused_attention.md:14]: As you said that TP should be the base consideration, I suggest that you only implement TP for gptj/llama without fused-attention in #361 (you can disable fused-attention with flag `--memory-f16` or compile and run on an unsupported platform (gcc<13; non-spr)). After that, I can implement fused-attention with TP.

**Review by airMeng — APPROVED** [review]

**Pull request merged**